docs/BackendAPI.md

```diff
@@ -0,0 +1,52 @@
+# ONNX Backend Interface Proposal
+
+We propose a cross-platform API for loading and executing ONNX graphs on optimized backends. High-level frameworks and applications can use this API to execute neural network and machine learning models. Hardware vendors can implement this API to expose specialized hardware accelerators and highly optimized software infrastructure to the users.
+
+## Core Features
+
+- Standardized interface for neural network inference on special-purpose accelerators, CPUs, GPUs, DSPs, and FPGAs
+- Based on widely supported technologies
+  - C API for function calls
+  - ONNX format for passing model graphs
+  - NCHW tensor layout for passing inputs and outputs
+- Dynamic discovery of available backends for model execution
+  - Multiple backends from different vendors can co-exist on the same system
+- Dynamic discovery of supported ONNX Operators on each backend
+
+### Optional Features:
+
+- Graphs with variable-shape inputs and/or outputs
+- Graphs with data-dependendent output shapes
+
+## How to Use ONNX Backend Interface
+
+0. (Optional) Use `onnx_backend_load` to dynamically load the ONNX Backend Interface library.
+1. Call `onnxGetNumBackends` to get the number of available backends. Note that it can be 0.
+2. Call `onnxGetBackendInfo` to check additional information about any available backend.
+3. Call `onnxGetBackendCompatibility` to check if your model, or parts of it, can run on the backend.
```

Comment:
What if only part it can run on BE ?
Who takes care of the graph partition ? 
The runnable part will still be dispatches to BE in the latter call ?

Reply:
If a backend supports only some operators of the graph, the user of the API (a deep learning framework) would have to do the partitioning. In the general case, the process is outlined below:
1. The user iterates operators in a model graph one-by-one, convert them to ONNX, and use `onnxGetBackendCompatibility` to check which of the operators can be offloaded to the backend.
2. The user constructs connected subgraphs of operators that can be offloaded to the backend.
3. The user queries the backend about it high-level performance characteristics using `ONNX_BACKEND_MACS_*` and `ONNX_BACKEND_MEMORY_BANDWIDTH` information queries. These data let the user build a simple roofline model of backend performance.
4. For every subgraph the user estimates time to do inference using the roofline model. The user additionally estimates time to transfer subgraph inputs to the backend using `ONNX_BACKEND_CPU_MEMORY_READ_BANDWIDTH` information query and to transfer subgraph outputs from the backend using `ONNX_BACKEND_CPU_MEMORY_WRITE_BANDWIDTH`.
5. If predicted time to transfer inputs to the backend, do inference, and transfer outputs from the backend is less than predicted time to do the inference on default engine (e.g. CPU), the user offloads the subgraph execution to the ONNX backend by calling `onnxSetGraphIO` and `onnxRunGraph`